packages/report-server/src/reportBuilder/transform/functions/groupRows/groupRows.ts

```diff
@@ -0,0 +1,120 @@
+
+type GroupRowsParams = {
+  createGroupKey: (row: Row) => string;
+  getMergeStrategy: (field: string) => keyof typeof mergeStrategies;
+  where: (parser: TransformParser) => boolean;
+};
+
+const paramsValidator = yup.object().shape({
+  by: starSingleOrMultipleColumnsValidator,
+  mergeUsing: yup.lazy((value: unknown) => {
+    const mergeStrategyNameValidator = yup
+      .mixed<keyof typeof mergeStrategies>()
+      .oneOf(Object.keys(mergeStrategies) as (keyof typeof mergeStrategies)[])
+      .required();
+    if (typeof value === 'string') {
+      return mergeStrategyNameValidator;
+    }
+
+    if (typeof value === 'object' && value !== null) {
+      const mergeStrategyMapValidator = Object.fromEntries(
+        Object.entries(value).map(([columnName]) => [columnName, mergeStrategyNameValidator]),
+      );
+      return yup.object().shape(mergeStrategyMapValidator);
+    }
+
+    throw new yup.ValidationError(
+      'mergeUsing must be either a single merge strategy, or a mapping between columns and merge strategies',
+    );
+  }),
+  where: yup.string(),
+});
+
+/**
+ * A group represents a key of fields, to all the values each row in that group had for the respective field
+ *
+ * eg.
+ * [{orgUnit: TO, BCD1: 4 }, {orgUnit: TO, BCD1: 7 }], groupBy orgUnit => group: { BCD1: [4, 7] }
+ */
+type Group = {
+  [fieldKey: string]: FieldValue[];
+};
+
+const buildGroups = (rows: Row[], params: GroupRowsParams) => {
+  const groupsByKey: Record<string, Group> = {};
+  const parser = new TransformParser(rows, functions);
+  const ungroupedRows: Row[] = []; // Rows that don't match the 'where' clause are left ungrouped
+
+  rows.forEach((row: Row) => {
+    if (!params.where(parser)) {
+      ungroupedRows.push(row);
+      parser.next();
+      return;
+    }
+    const groupKey = params.createGroupKey(row);
+    groupsByKey[groupKey] = addRowToGroup(groupsByKey[groupKey] || {}, row);
+    parser.next();
+  });
+
+  return { groups: Object.values(groupsByKey), ungroupedRows };
+};
+
+const addRowToGroup = (group: Group, row: Row): Group => {
+  const newGroup = { ...group };
+  Object.keys(row).forEach((field: string) => {
+    newGroup[field] = [...(newGroup[field] || []), row[field]];
+  });
+  return newGroup;
+};
+
+const mergeGroups = (groups: Group[], params: GroupRowsParams): Row[] => {
+  return groups.map(group => {
+    const mergedRow: Row = {};
+    Object.entries(group).forEach(([fieldKey, fieldValue]) => {
+      const mergeStrategy = params.getMergeStrategy(fieldKey);
+      const mergedValue = mergeStrategies[mergeStrategy](fieldValue);
+      if (mergedValue !== undefined) {
+        mergedRow[fieldKey] = mergedValue;
+      }
+    });
+    return mergedRow;
+  });
+};
+
+const groupRows = (rows: Row[], params: GroupRowsParams): Row[] => {
+  const { groups, ungroupedRows } = buildGroups(rows, params);
+  const mergedRows = mergeGroups(groups, params);
+  return mergedRows.concat(ungroupedRows);
```

Comment:
Re-wrote the functions in this class to be a bit cleaner. In doing so it's really solidified what's going on in this operation which is we are first `group`ing rows, and then `merging` those groups.

This ties into a bit of what @edmofro was saying that perhaps a better term for this operation would be `mergeRows`. It also relates to this comment here: https://beyond-essential.slab.com/posts/report-server-transform-config-design-change-proposals-k8l3tch0#thread-12c3fjiw

Would appreciate the reviewer's thoughts on this. Some possible options:
1. Keep as is, `groupRows` `by` `mergeUsing`
2. Use single term, either `groupRows` `by` `groupUsing` or `mergeRows` `by` `mergeUsing`
3. Use 2 terms, either `groupAndMergeRows` `groupBy` `mergeUsing` or just `mergeRows` `groupBy` `using` ?

Reply:
Ah thanks for looking up the R style! I'm helping Gerry edit a viz via viz-builder today, so I might get his opinion on merge vs group!

Just so ya know, @Alastair-L is also in the `mergeRows` > `groupRows` camp, so you guys are picking up steam!

Reply:
Haha yay @Alastair-L 🙌 

I know the `withMatching` was a sticking point, and I had a think about some alternatives if you are still workshopping this, combinations of which are below:

```
[
  {
    "transform": "mergeRows",
    "withSame": "period",
    "using": "sum"
  },
  {
    "transform": "combineRows",
    "using": "sum",
    "splitOn": "period"
  },
  {
    "transform": "squashRows",
    "onePer": "period",
    "using": "sum"
  },
  {
    "transform": "mergeRows",
    "groupBy": "period",
    "using": "sum"
  }
]
```
I think the first and last (your suggestion) are my faves, and I lean slightly towards the last

Reply:
So when I ran through this with @gerrybes he definitely described the operation as 'grouping by'. However, once I showed him the merge config he also felt it made total sense 😂

Are we all okay with `mergeRows` `groupBy` `using` ? It seems people really like group by as a term, but merge rows is a better description of the actual operation of this transform?